src/backend/distributed/planner/insert_select_planner.c

```diff
@@ -0,0 +1,234 @@
+{
+	List *topLevelCTEList;
+	int topLevelsUp;
+};
+
+
+static DeferredErrorMessage * ErrorIfCoordinatorInsertSelectUnsupported(
+	Query *insertSelectQuery);
+static bool DecrementCTELevelsWalker(Node *node,
+									 struct DecrementCTELevelsContext *context);
+
+
+/*
+ * CreatteCoordinatorInsertSelectPlan creates a query plan for a SELECT into a
+ * distributed table.
+ */
+MultiPlan *
+CreateCoordinatorInsertSelectPlan(Query *parse)
+{
+	Query *insertSelectQuery = copyObject(parse);
+
+	RangeTblEntry *subqueryRte = ExtractSelectRangeTableEntry(insertSelectQuery);
+	RangeTblEntry *insertRte = ExtractInsertRangeTableEntry(insertSelectQuery);
+	Oid targetRelationId = insertRte->relid;
+
+	Query *subquery = (Query *) subqueryRte->subquery;
+
+	ListCell *selectTargetCell = NULL;
+	ListCell *insertTargetCell = NULL;
+
+	MultiPlan *multiPlan = CitusMakeNode(MultiPlan);
+	multiPlan->operation = CMD_INSERT;
+
+	multiPlan->planningError =
+		ErrorIfCoordinatorInsertSelectUnsupported(insertSelectQuery);
+
+	if (multiPlan->planningError != NULL)
+	{
+		return multiPlan;
+	}
+
+	if (list_length(insertSelectQuery->cteList) > 0)
+	{
+		struct DecrementCTELevelsContext context;
+		List *combinedCTEList = NIL;
+
+		/* prepend top-level CTEs in subquery */
+		combinedCTEList = list_concat(combinedCTEList, insertSelectQuery->cteList);
+		combinedCTEList = list_concat(combinedCTEList, subquery->cteList);
+		subquery->cteList = combinedCTEList;
+
+		context.topLevelCTEList = insertSelectQuery->cteList;
+		context.topLevelsUp = 0;
+
+		/* correct ctelevelsup for range table entries that refer to top-level CTEs */
+		DecrementCTELevelsWalker((Node *) subquery, &context);
+	}
+
+	ReorderInsertSelectTargetLists(insertSelectQuery, insertRte, subqueryRte);
+
+	/* add casts when the SELECT output does not directly match the table */
+	forboth(insertTargetCell, insertSelectQuery->targetList,
+			selectTargetCell, subquery->targetList)
+	{
+		TargetEntry *insertTargetEntry = (TargetEntry *) lfirst(insertTargetCell);
+		TargetEntry *selectTargetEntry = (TargetEntry *) lfirst(selectTargetCell);
+
+		Var *columnVar = NULL;
+		Oid columnType = InvalidOid;
+		int32 columnTypeMod = 0;
+		Oid selectOutputType = InvalidOid;
+
+		/* indirection is not supported, e.g. INSERT INTO table (composite_column.x) */
+		if (!IsA(insertTargetEntry->expr, Var))
+		{
+			ereport(ERROR, (errmsg("can only handle regular columns in the target "
+								   "list")));
+		}
+
+		columnVar = (Var *) insertTargetEntry->expr;
+		columnType = get_atttype(targetRelationId, columnVar->varattno);
+		columnTypeMod = get_atttypmod(targetRelationId, columnVar->varattno);
+		selectOutputType = columnVar->vartype;
+
+		/*
+		 * If the type in the target list does not match the type of the column,
+		 * we need to cast to the column type. PostgreSQL would do this
+		 * automatically during the insert, but we're passing the SELECT
+		 * output directly to COPY.
+		 */
+		if (columnType != selectOutputType)
+		{
```

Comment:
Minor: Maybe move `TargetEntry *selectTargetEntry = (TargetEntry *) lfirst(selectTargetCell);` to here? That'd seem a bit confusing but probably more convinient wrt the conventions?